### ISSUE TITLE
SELECT with bitmap index get wrong data (or Assert Fail). 

### ISSUE THREAD

**Issue by linwenbo** [greenplum-db/gpdb#13446]
### Greenplum version or build
Master branch, commit 73e50fbaf1578be81522f2d1eaf70f5905554d5a

### start GPDB demo cluster
source Release/greenplum_path.sh
make create-demo-cluster
source gpAux/gpdemo/gpdemo-env.sh

### Step to reproduce the behavior

```sql
gpadmin@zlyu:~$ cat a.sql
create table test(a int4, b int4);
create index idx on test using bitmap(b);

insert into test values (generate_series(1, 40000), 1);

delete from test;

insert into test values (generate_series(1, 40000), 1);

set enable_seqscan = off;
set enable_bitmapscan = off;

select count(*) from test where b = 1;

gpadmin@zlyu:~$ psql -f a.sql
psql:a.sql:1: NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
CREATE INDEX
INSERT 0 40000
DELETE 40000
INSERT 0 40000
SET
SET
psql:a.sql:13: ERROR:  Unexpected internal error (assert.c:44)  (seg0 slice1 127.0.1.1:7002 pid=36803) (assert.c:44)
DETAIL:  FailedAssertion("!(words->firstTid == result->nextTid)", File: "bitmaputil.c", Line: 265)
```

**Comment by linwenbo** [comment]
Here is some additional information. I don't know whether it's useful。

I move the bitmap index code to postgresql V12, run some SQL as above, and I got different **select count(*)** result  with condition and without condition. I try to debug it and I found this:
with the condition such as **select count(*) from test where b = 1**, I got a weird nextTid in _bitmap_findnexttid: when the BM_MAX_TUPLES_PER_PAGE is 65536 on my machine. I got 4784128 for nextTid, and 4784128%65536 is 0, which means the BM_INT_GET_OFFSET will get 65536 on this nextTid, but the offset number is an uint16 type number, this will casue a overflow on offset number。
but when I run the same select without the condition, I can't get any value which mod 65536 is 0.

I didn't make any change on the implement of bitmap index. guess the problem is in _bitmap_findnexttid function.

**Comment by SmartKeyerror** [comment]
RCA: This issue is bringed in by **#11377**, which resolved data corrupted with concurrent insert update full bitmap page, the culprit is the function `_bitmap_catchup_to_next_tid`, which catch up to the nextTid we need to check from last iteration. This will work fine if there has actually concurrent insert update full bitmap page happend, but it forget during the bitmap index scan, `BMIterateResult`  can only save up to 16*1024 (**`BM_BATCH_TIDS`**) tids, so if a bitmap page stores more than 16384 bits, this function `_bitmap_catchup_to_next_tid` will cause wrong bitmap index scan. 

We can use a simple example to reproduce it:

```
postgres=# create table s(a int4, b int4);
postgres=# create index idx on s using bitmap(b);

postgres=# insert into s select 1, 1 from generate_series(0, 16384);

postgres=# select count(*) from s where b = 1;
 count 
-------
 16385
(1 row)

postgres=# set optimizer = false;
postgres=# set enable_seqscan = off;
postgres=# set enable_bitmapscan = off;

postgres=# select count(*) from s where b = 1;
 count 
-------
 16384
(1 row)
```

As we can see, we insert 16385 rows in s, and all of the data will be srored in one segment. Therefore, it is guaranteed that when the executor obtains the tids in the bitmap page, it must execute 2 times to get all the data.

That is to say, `_bitmap_catchup_to_next_tid` will be executed twice, and in the second execution, the wrong offset will be obtained, resulting in an incorrect final result.

**Comment by linwenbo** [comment]
> RCA: This issue is bringed in by **#11377**, which resolved data corrupted with concurrent insert update full bitmap page, the culprit is the function `_bitmap_catchup_to_next_tid`, which catch up to the nextTid we need to check from last iteration. This will work fine if there has actually concurrent insert update full bitmap page happend, but it forget during the bitmap index scan, `BMIterateResult` can only save up to 16*1024 (**`BM_BATCH_TIDS`**) tids, so if a bitmap page stores more than 16384 bits, this function `_bitmap_catchup_to_next_tid` will cause wrong bitmap index scan.
> 
> We can use a simple example to reproduce it:
> 
> ```
> postgres=# set optimizer = false;
> postgres=# set enable_seqscan = off;
> postgres=# set enable_bitmapscan = off;
> 
> postgres=# create table s(a int4, b int4);
> postgres=# create index idx on s using bitmap(b);
> 
> # this will get correct result
> postgres=#insert into s select 1, 1 from generate_series(0, 16383);
> postgres=# select count(*) from s where b = 1;
>  count 
> -------
>  16384
> (1 row)
> 
> postgres=# truncate table s;
> 
> # but if we insert more than 136384 rows, it will fail, the result should be 13685
> postgres=# insert into s select 1, 1 from generate_series(0, 16384);
> postgres=# select count(*) from s where b = 1;
>  count 
> -------
>  16384
> (1 row)
> ```
> 
> As we can see, we insert 16385 rows in s, and all of the data will be srored in one segment. Therefore, it is guaranteed that when the executor obtains the tids in the bitmap page, it must execute 2 times to get all the data.
> 
> That is to say, `_bitmap_catchup_to_next_tid` will be executed twice, and in the second execution, the wrong offset will be obtained, resulting in an incorrect final result.

But why the **SELECT** without **where b = 1** will get the correct result? the stack of select with or without the condition both will finally call bmgettuple function to get the TID of heap tuple. the only difference is scankey.

**Comment by linwenbo** [comment]
I found the _bitmap_findbitmaps function in bitmapsearch.c will add a item in lovItemPoss which the blockno is BM_LOV_STARTPAGE and the offset is 1 when the numberOfKeys is 0 which means there is no scanKey.
```
		/* When there are no scan keys, all bitmap vectors are included,
		 * including the one for the NULL value.
		 * This can happen when the index is created with predicates.
		 */
		if (scan->numberOfKeys == 0)
		{
			ItemPos *itemPos;
			itemPos = (ItemPos*)palloc0(sizeof(ItemPos));
			itemPos->blockNo = BM_LOV_STARTPAGE;;
			itemPos->offset = 1;
			lovItemPoss = lappend(lovItemPoss, itemPos);

			scanPos->nvec++;
		}
```
I try to make the if condition allways true. and then the **select** with **where b = 1**  returns a correct result.

**Comment by SmartKeyerror** [comment]
> But why the **SELECT** without **where b = 1** will get the correct result? the stack of select with or without the condition both will finally call bmgettuple function to get the TID of heap tuple. the only difference is scankey.

Not exactly, select without where clauses will use different query plan, although they will both inkove `bmgettuple`, but
they will use different code paths:

```
postgres=# explain (analyze, costs off) select count(*) from s;
                                             QUERY PLAN                                             
----------------------------------------------------------------------------------------------------
 Finalize Aggregate (actual time=2705.447..2705.467 rows=1 loops=1)
   ->  Gather Motion 3:1  (slice1; segments: 3) (actual time=0.419..2705.418 rows=3 loops=1)
         ->  Partial Aggregate (actual time=0.050..0.090 rows=1 loops=1)
               ->  Index Only Scan using idx on s (actual time=47.100..2634.684 rows=16385 loops=1)
                     Heap Fetches: 16385

postgres=# explain (analyze, costs off) select count(*) from s where b = 1;
                                          QUERY PLAN                                           
-----------------------------------------------------------------------------------------------
 Finalize Aggregate (actual time=2063.554..2063.570 rows=1 loops=1)
   ->  Gather Motion 3:1  (slice1; segments: 3) (actual time=0.380..2063.526 rows=3 loops=1)
         ->  Partial Aggregate (actual time=0.052..0.065 rows=1 loops=1)
               ->  Index Scan using idx on s (actual time=46.640..2005.816 rows=16384 loops=1)
                     Index Cond: (b = 1)
```

**Comment by linwenbo** [comment]
> Not exactly, select without where clauses will use different query plan, although they will both inkove bmgettuple
yes, the query plan is different. but the point of my confusion is if the second execution of `_bitmap_catchup_to_next_tid ` cause the wrong offset,  then whatever the query path is , they finally will execute `_bitmap_catchup_to_next_tid` twice and lead to a wrong result.  but the truth is when execute the query without the where clause, we will get a corrent result。

**Comment by SmartKeyerror** [comment]
I just give a runtime explanation on this, the original design intent is difficult to illustrate in a few words, we can use a gdb to attach the function of `_bitmap_findbitmaps()` and `_bitmap_next()` when we execute below SQL:

```
select count(*) from b;
```

And after finished `_bitmap_findbitmaps()`, the number of related bitmap vectors will 2, as `scanPos->nvec = 2`, and this will lead function `_bitmap_next()` has distinct behavior from queries with where clauses: it'll update the `words->firstTis` equal to `result->nextTid`, this will make the function `_bitmap_catchup_to_next_tid` return directly, `result->lastScanWordNo` will be correct, so the reult is correct. 

Hope this can clear your confusion.

**Comment by linwenbo** [comment]
I made some change to fix this issue:
[https://github.com/linwenbo/gpdb/commit/6365f54b4efc87861cf509923f8755f76a9f0b70](url)
I found the catchup function changed the words which will be used to get the next result->nextTids. so I try to make the catchup function really to catchup the result->nextTid without changing the words->nwords and result->lastScanWordNo. ensure to get correct result in the next iteration in `_bitmap_findnexttids`.
But it seems yours is better.

**Comment by interma** [comment]
> > But why the **SELECT** without **where b = 1** will get the correct result? the stack of select with or without the condition both will finally call bmgettuple function to get the TID of heap tuple. the only difference is scankey.
> 
> Not exactly, select without where clauses will use different query plan, although they will both inkove `bmgettuple`, but they will use different code paths:
> 
> ```
> postgres=# explain (analyze, costs off) select count(*) from s;
>                                              QUERY PLAN                                             
> ----------------------------------------------------------------------------------------------------
>  Finalize Aggregate (actual time=2705.447..2705.467 rows=1 loops=1)
>    ->  Gather Motion 3:1  (slice1; segments: 3) (actual time=0.419..2705.418 rows=3 loops=1)
>          ->  Partial Aggregate (actual time=0.050..0.090 rows=1 loops=1)
>                ->  Index Only Scan using idx on s (actual time=47.100..2634.684 rows=16385 loops=1)
>                      Heap Fetches: 16385
> 
> postgres=# explain (analyze, costs off) select count(*) from s where b = 1;
>                                           QUERY PLAN                                           
> -----------------------------------------------------------------------------------------------
>  Finalize Aggregate (actual time=2063.554..2063.570 rows=1 loops=1)
>    ->  Gather Motion 3:1  (slice1; segments: 3) (actual time=0.380..2063.526 rows=3 loops=1)
>          ->  Partial Aggregate (actual time=0.052..0.065 rows=1 loops=1)
>                ->  Index Scan using idx on s (actual time=46.640..2005.816 rows=16384 loops=1)
>                      Index Cond: (b = 1)
> ```

curious about the first plan: I remember bitmap index doesn't support index only scan 

https://github.com/greenplum-db/gpdb/blob/f89ccb3a5734a5460f29464391b3ecdd46340468/src/backend/access/bitmap/bitmapsearch.c#L64

**Comment by SmartKeyerror** [comment]
This issue has been fixed by #13479 and #13522, so close this one.
